### PR TITLE
unit test for flux_dataset dataloader checkpointing

### PR DESCRIPTION
**Pull request by wesleytruong**
Adds a unit test for loading flux dataset from a checkpoint. Creates the new dataloader from a checkpoint then ensures that the next generated labels and tokens are the same in both dataloaders, starting from the checkpoint.

### CODE REVIEW

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 1]
```diff
@@ -97,3 +99,63 @@ def _build_dataloader(
...
+        for world_size in [2, 4]:
+            for rank in range(world_size):
+                batch_size = 4
+
+                path = "torchtitan.experiments.flux.job_config"
+                config_manager = ConfigManager()
+                config = config_manager.parse_args(
+                    [
+                        f"--experimental.custom_args_module={path}",
+                        # Profiling options
```
Comment (wwwjn) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: If these options are not needed, pls delete them, I forgot to clean them in `_test_flux_dataloader`

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 2]
```diff
@@ -97,3 +99,63 @@ def _build_dataloader(
...
+                        "--training.img_size",
+                        str(256),
+                        "--training.dataset",
+                        dataset_name,
+                        "--training.local_batch_size",
+                        str(batch_size),
+                        "--training.seed",
+                        "0",
+                        "--training.classifer_free_guidance_prob",
+                        "0.447",
```
Comment (wwwjn) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:38]: Why enabling classifer-free guidance here?  Would it cause random dropout in the dataloader?

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 3]
```diff
@@ -97,3 +99,63 @@ def _build_dataloader(
...
+
+                # Create new dataloader, restore checkpoint, and check if next data yielded is the same as above
+                dl_resumed = self._build_dataloader(config, world_size, rank)
+                dl_resumed.load_state_dict(state)
+                it_resumed = iter(dl_resumed)
+
+                for i in range(50):
+                    torch.manual_seed(i)
+                    expected_input_ids, expected_labels = next(it)
+                    torch.manual_seed(i)
```
Comment (wwwjn) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py:92]: You could add some NOTE here, saying that we need to manually set random seed every time before calling next() because the randomness in dataloader behaviour.

**Review by wwwjn — COMMENTED** [review]
Nice work! Please fix lint as well

**Comment by wesleytruong** [comment]
Fixed linting errors and comments

> Why enabling classifer-free guidance here? Would it cause random dropout in the dataloader?

I removed this, it was a mistake from copy-pasting. It doesn't have an effect on the test due to the torch.manual_seed but still is unnecessary

@wwwjn

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 4]
```diff
@@ -16,6 +18,7 @@ class TestFluxDataLoader:
     def test_load_dataset(self):
         for dataset_name in ["cc12m-test"]:
             self._test_flux_dataloader(dataset_name)
+            self._test_flux_dataloader_checkpointing(dataset_name)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: With the new one, the old one's value is not clear.
I suggest we merge the two into one.
Also there is no need for this for-loop and extra layer of `test_load_dataset` -> `_test_flux_dataloader`, we can do the test just in one function.

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 5]
```diff
@@ -97,3 +93,55 @@ def _build_dataloader(
             tokenizer=None,
             infinite=True,
         )
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: `_build_dataloader` on line 87 is not very useful, we can just directly use `build_flux_dataloader`

**Comment by wesleytruong** [comment]
Refactored the two tests according to @tianyu-l
The test now checks for the correct tensor shapes during the first num_steps, then for the next num_steps it ensures the loaded from checkpoint dataloader generates tokens and labels correctly
```build_dataloader``` is also called directly now instead of through ```_build_dataloader```

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 6]
```diff
@@ -14,86 +16,99 @@
...
+                        "--encoder.max_t5_encoding_len",
+                        "512",
+                    ]
+                )
+
+                with maybe_enable_profiling(
+                    config, global_step=0
+                ) as torch_profiler, maybe_enable_memory_snapshot(
+                    config, global_step=0
+                ) as memory_profiler:
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: I think we can remove these code related to profiling.

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 7]
```diff
@@ -14,86 +16,99 @@
...
+                        # across dataloaders for testing purposes.
+                        torch.manual_seed(i)
+                        expected_input_ids, expected_labels = next(it)
+                        torch.manual_seed(i)
+                        input_ids, labels = next(it_resumed)
+
+                        if torch_profiler:
+                            torch_profiler.step()
+                        if memory_profiler:
+                            memory_profiler.step()
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: remove

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 8]
```diff
@@ -14,86 +16,99 @@
...
+                        )
+                        assert torch.equal(
+                            input_ids["t5_tokens"], expected_input_ids["t5_tokens"]
+                        )
+                        assert torch.equal(labels, expected_labels)
+
+                    if torch_profiler:
+                        torch_profiler.step()
+                    if memory_profiler:
+                        memory_profiler.step(exit_ctx=True)
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: remove

**Review by tianyu-l — COMMENTED** [review]
Looks better. I had one more comment on removing profiling code, as they are not quite related to the test itself.

> The test now checks for the correct tensor shapes during the first num_steps, then for the next num_steps it ensures the loaded from checkpoint dataloader generates tokens and labels correctly

This is helpful. Let's put them as comments in the test itself, so that people who look at the test later can easily understand what you are doing here.

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 9]
```diff
@@ -14,86 +16,99 @@
...
+                        "--training.dataset",
+                        dataset_name,
+                        "--training.local_batch_size",
+                        str(batch_size),
+                        "--training.seed",
+                        "0",
+                        "--training.classifer_free_guidance_prob",
+                        "0.447",
+                        "--encoder.t5_encoder",
+                        "google/t5-v1_1-small",
```
Comment (wwwjn) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: You could change this line to "google/t5-v1_1-xxl", which is used in our actual FLUX training.

File: torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py [hunk 10]
```diff
@@ -14,86 +16,99 @@
...
+                    for i in range(0, num_steps):
+                        input_data, labels = next(it)
+                        if torch_profiler:
+                            torch_profiler.step()
+                        if memory_profiler:
+                            memory_profiler.step()
+
+                        assert len(input_data) == 2  # (clip_encodings, t5_encodings)
+                        assert labels.shape == (batch_size, 3, 256, 256)
+                        # assert input_data["clip_tokens"].shape[0] == batch_size
```
Comment (wwwjn) [inline torchtitan/experiments/flux/tests/unit_tests/test_flux_dataloader.py]: Here's the right shape of tensor

assert input_data["clip_tokens"].shape[0] == batch_size   # (bsz, 1, 77) , 7
assert input_data["t5_tokens"].shape == (batch_size, 1, 256)  # (bsz, 1, seq_length)

**Comment by wesleytruong** [comment]
Removed the memory profiling, and added the unit test for tensor dimensions. @tianyu-l 
Added unit the test for the tensor shape after tokenization, but used a different value for the test. I believe this is the correct shape since it matches the shape of a tokenized sequence. @wwwjn
```
                    assert input_data["clip_tokens"].shape == (
                        batch_size,
                        1,
                        77,
                    )
                    assert input_data["t5_tokens"].shape == (
                        batch_size,
                        1,
                        256,
                    )
```

**Review by tianyu-l — APPROVED** [review]
Looks great, thank you!
